javascript/stimulus_reflex.js

```diff
@@ -1,166 +1,75 @@
-              reflexId
-            )
-            registerReflex(reflexData)
-          }
-
-          if (reflexes[reflexId]) {
-            reflexes[reflexId].totalOperations = totalOperations
-            reflexes[reflexId].pendingOperations = totalOperations
-            reflexes[reflexId].completedOperations = 0
-            CableReady.perform(reflexOperations)
-          }
-        }
+class StimulusReflexController extends Controller {
+  constructor (...args) {
+    super(...args)
+    register(this)
+  }
+}
 
-        // run piggy back operations after stimulus reflex behavior
-        CableReady.perform(data.operations)
-      },
-      connected: () => {
-        actionCableSubscriptionActive = true
-        emitEvent('stimulus-reflex:connected')
-      },
-      rejected: () => {
-        actionCableSubscriptionActive = false
-        emitEvent('stimulus-reflex:rejected')
-        if (Debug.enabled) console.warn('Channel subscription was rejected.')
-      },
-      disconnected: willAttemptReconnect => {
-        actionCableSubscriptionActive = false
-        emitEvent('stimulus-reflex:disconnected', willAttemptReconnect)
-      }
-    })
+// Initializes StimulusReflex by registering the default Stimulus controller with the passed Stimulus application.
+//
+// - application - the Stimulus application
+// - options
+//   * controller - [optional] the default StimulusReflexController
+//   * consumer - [optional] the ActionCable consumer
+//   * debug - [false] log all Reflexes to the console
+//   * params - [{}] key/value parameters to send during channel subscription
+//   * isolate - [false] restrict Reflex playback to the tab which initiated it
+//
+const initialize = (application, initializeOptions = {}) => {
+  const { controller, consumer, debug, params, isolate } = initializeOptions
+  actionCable.consumer = consumer
+  actionCable.params = params
+  isolationMode.set(!!isolate)
+  reflexes.app = application
+  reflexes.app.schema = { ...defaultSchema, ...application.schema }
+  reflexes.app.register(
+    'stimulus-reflex',
+    controller || StimulusReflexController
+  )
+  Debug.set(!!debug)
+  const observer = new MutationObserver(setupDeclarativeReflexes)
+  observer.observe(document.documentElement, {
+    attributeFilter: [reflexes.app.schema.reflexAttribute],
+    childList: true,
+    subtree: true
+  })
 }
 
-// Extends a regular Stimulus controller with StimulusReflex behavior.
+// Registers a Stimulus controller and extends it with StimulusReflex behavior
 //
-// Methods added to the Stimulus controller:
-// - stimulate
-// - __perform
+// controller - the Stimulus controller
+// options - [optional] configuration
 //
-const extendStimulusController = controller => {
+const register = (controller, options = {}) => {
+  const channel = 'StimulusReflex::Channel'
+  controller.StimulusReflex = { ...options, channel }
+  actionCable.createSubscription(controller)
   Object.assign(controller, {
```

Comment:
There's possibly one more improvement that could be made here. You could move the object that you assign here into a named object, similar to what has been done elsewhere in the project.

Reply:
You'll have to forgive me, it's been over a month. Could you give me a reference to an example of the style/implementation that you're referring to so I can get my brain back in this timeline?

Reply:
To anyone watching: Jonathan and I back-channeled and we're on the same page, but what he's suggesting will come in later PRs which are more specifically/themeatically about removing AC-specific code from the main event loop.